server/util/db/db.go

```diff
@@ -650,17 +652,52 @@ func GetConfiguredDatabase(env environment.Env) (interfaces.DBHandle, error) {
 	}
 	go statsRecorder.poll()
 
-	if *autoMigrateDBAndExit {
-		if err := runMigrations(driverName, primaryDB); err != nil {
-			log.Fatalf("Database auto-migration failed: %s", err)
+	if *autoMigrateDB || *autoMigrateDBAndExit {
+		sqlStrings := make([]string, 0)
+		if *autoMigrateDBOutputFile != "" {
+			if err := primaryDB.Callback().Raw().Register("save_sql", func(db *gorm.DB) {
+				executedSql := db.Statement.SQL.String()
+				if !strings.HasPrefix(executedSql, "SELECT") {
+					sqlStrings = append(sqlStrings, executedSql)
+				}
+			}); err != nil {
+				return nil, err
+			}
 		}
-		log.Infof("Database migration completed. Exiting due to --auto_migrate_db_and_exit.")
-		os.Exit(0)
-	}
-	if *autoMigrateDB {
+
 		if err := runMigrations(driverName, primaryDB); err != nil {
 			return nil, err
 		}
+
+		if *autoMigrateDBOutputFile != "" {
```

Comment:
WDYT about just printing the changes to stdout below some marker line?

Like:

"Schema changes follow:"
"CREATE TABLAH"
"INSERT TABLAH"
"ALTER TABLAH"

or

"No schema changes necessary."

Then this code becomes simpler (yay) and your calling script can just check for the presence of "No schema changes necessary." and bail if it doesn't see that. Or alternatively, later, it could include the schema changes in a message to us saying please fix or whatever.

Reply:
I was thinking it would be nice if the bash script could easily parse the schema changes, so it could do something like send a slack message with the schema changes, and people could use a slackbot to still continue the auto-deploy depending on the changes

We could still do this if we just printed the changes to stdout, but then the bash script would have to do some additional (likely gross) regex parsing to extract the schema changes. So slight preference for the output file because I don't think it adds too much complexity, but don't feel strongly if you do

Reply:
Ack -- I understand what you're saying here. One thing that's nice about printing this stuff to stdout is that then it's easier to compose this command in bash and you don't need to worry about figuring out a file path. It's more unix-y if that makes sense. For instance, you could do:

```
SCHEMA_DIFF = $(/path/to/buildbuddy --print_schema_diff_and_exit)
if [[ $SCHEMA_DIFF ]]; then
  echo "no changes yay"
else
  slack_send $SCHEMA_DIFF
fi
```

I suggested printing lines before the schema change in my previous comment because I thought we might be printing stuff to stdout and needed a way to distinguish, but I checked and logging etc is all going to stderr, so you can just print the schema to sdout and then exit. So the following would dump the schema to just a file without log changes:

```
/path/to/buildbuddy --print_schema_diff_and_exit > schema_changes.diff 2>/dev/null
```

The other thing is that it's *nice* to have all the shit code be in bash -- it's sort of what bash is meant for :)

Reply:
Good point about logs going to stderr - that resolves my earlier concerns! Changed to print to stdout